corehq/pillows/mappings/utils.py

```diff
@@ -11,3 +12,63 @@ def mapping_from_json(filename):
                 .replace('"__DATE_FORMATS_ARR__"', json.dumps(DATE_FORMATS_ARR)))
         mapping = json.loads(data)
     return mapping
+
+
+def transform_for_es7(original_mapping):
+    if settings.ELASTICSEARCH_MAJOR_VERSION != 7:
+        # v1 and v2 have same mapping
+        return original_mapping
+    mapping = original_mapping.copy()
+    mapping = _transform_types(mapping)
+    if "_all" in mapping:
+        mapping.pop("_all")
+    return mapping
+
+
+def _transform_types(mapping):
+    """
+    Replaces mapping types to be compatible with Elasticsearch version 7
+
+    String fileds are text/keyword depending on analyzed or not
+        {"type": "string", "index", "analyzed", **kw} --> {"type": "text", **kw}
+        {"type": "string", "index", "not_analyzed", **kw} --> {"type": "keyword", **kw}
+        {"type": "string", **kw} --> {"type": "text", **kw}
+
+    multi_field is dropped since 'fields' attrib captures the type
+        {"type": "multi_field", **kw} --> {"type": "text", **kw}
+
+    '_all' and 'include_in_all' are dropped
+
+    """
+    if settings.ELASTICSEARCH_MAJOR_VERSION != 7:
+        # v1 and v2 have same mapping
+        return mapping
+
+    if isinstance(mapping, dict):
```

Comment:
The idea is to mutate it recursively. Also It's a private function called from just `transform_for_es7`.

Reply:
I see, and `transform_for_es7` makes a copy before sending it in here.

Previously [I advised you to remove a similar _copy_ call](https://github.com/dimagi/commcare-hq/pull/28267#discussion_r463839625), but I am now regretting that suggestion since it did not adhere to the best practice. The best practice is: _functions that return a value should not mutate any of their arguments_.

Reply:
Haha, exactly, generally I use copy whenever it's mutable arg.